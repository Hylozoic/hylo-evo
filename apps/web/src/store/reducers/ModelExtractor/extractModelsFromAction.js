import { castArray } from 'lodash'
import ModelExtractor from 'store/reducers/ModelExtractor'

// extract immediately instead of creating new actions
export default function extractModelsFromAction (action, session) {
  getRoots(action).forEach(({ payload, modelName, append }) =>
    ModelExtractor.addAll({ session, root: payload, modelName, append }))
}

export function getRoots (action) {
  const { meta: { extractModel }, payload } = action
  const keys = Object.keys(payload.data)
  const roots = []

  if (typeof extractModel === 'string') {
    // shorthand syntax: payload.data must have exactly one child, and the
    // value of extractModel is the model name
    if (keys.length !== 1) {
      throw new Error(`Bad data for ${action.type}: payload.data should have exactly one child`)
    }

    roots.push({
      payload: payload.data[keys[0]],
      modelName: extractModel,
      append: true
    })
  } else {
    castArray(extractModel).forEach(config => {
      roots.push({
        payload: config.getRoot(payload.data),
        modelName: config.modelName,
        append: config.append
      })
    })
  }

  return roots
}
