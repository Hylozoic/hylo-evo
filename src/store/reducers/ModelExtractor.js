import { Attribute, ForeignKey, ManyToMany } from 'redux-orm/lib/fields'
import { compact, filter, mapValues } from 'lodash'
import { isUndefined, omitBy } from 'lodash/fp'

export default class ModelExtractor {
  static addAll ({ session, root, modelName }) {
    const extractor = new ModelExtractor(session)
    extractor.walk(root, modelName)
    extractor.addAll()
  }

  constructor (session) {
    this.session = session
    this.accumulator = []
  }

  addAll () {
    this.mergedNodes().forEach(({ modelName, payload }) => {
      const model = this.session[modelName]
      model.hasId(payload.id)
        ? model.withId(payload.id).updateAppending(payload)
        : model.create(payload)
    })
  }

  walk (node, modelName) {
    if (Array.isArray(node)) {
      return node.forEach(x => this.walk(x, modelName))
    }

    const model = this.session[modelName]

    const normalized = omitBy(isUndefined, mapValues(node, (value, key) => {
      var type = model.fields[key]

      if (type instanceof Attribute) {
        return value
      }

      if (type instanceof ForeignKey) {
        return this._walkOne(value, type)
      }

      if (type instanceof ManyToMany) {
        return this._walkMany(value, type)
      }

      if (!type && key in model.prototype) {
        // this is a reverse relation defined by relatedName
        type = model.virtualFields[key]

        if (type instanceof ForeignKey) {
          // each of the related values needs to have a foreign key back to
          // the current value...
          this._walkMany(value, type, {[type.relatedName]: node.id})

          // ...and because the related values store the foreign key, the
          // current value does not need to record anything about the relation,
          // so we return nothing
          return
        }

        if (type instanceof ManyToMany) {
          return this._walkMany(value, type)
        }
      }

      if (!type) return value
      throw new Error(`don't know how to handle type: ${type}`)
    }))

    if (normalized.id) {
      this.accumulator.push({
        modelName: model.modelName,
        payload: normalized
      })
    }
  }

  mergedNodes () {
    return mergeDuplicates(this.accumulator)
  }

  _walkOne (value, type) {
    if (typeof value !== 'object') return value
    this.walk(value, type.toModelName)
    return value.id
  }

  _walkMany (value, type, extraProps) {
    const items = Array.isArray(value) ? value : value[QUERY_SET_ITEMS_KEY]
    return items.map(x => {
      this.walk(extraProps ? Object.assign(x, extraProps) : x, type.toModelName)
      return x.id
    })
  }
}

// This is more complicated than a simple groupBy(modelName + id).map(merge)
// because we need to preserve the order of nodes.
function mergeDuplicates (nodes) {
  const usedIndexes = {}

  // go through all nodes we accumulated from beginning to end and merge
  // duplicates for each one.
  return compact(nodes.map((node, index) => {
    // skip this node if it's already been merged with an earlier one
    if (usedIndexes[index]) return
    const { modelName, payload: { id } } = node

    // find nodes later in the list that have the same model name and ID,
    // collect them, and mark them to be skipped later
    const duplicates = filter(nodes, (other, index2) => {
      if (index2 > index &&
        other.modelName === modelName && other.payload.id === id) {
        usedIndexes[index2] = true
        return true
      }
    })

    if (duplicates.length === 0) return node

    // merge all the payloads into a new one.
    // we don't use lodash.merge here because it merges arrays by
    // concatenating them, and we don't need its "deep" behavior since these
    // nodes are already normalized.
    const payloads = [{}, node.payload].concat(duplicates.map(d => d.payload))
    return {
      modelName,
      payload: Object.assign.apply(null, payloads)
    }
  }))
}

// this is the key under which our GraphQL API stores the list of items for
// "query-set style" pagination, e.g.:
//
//   person(id: 2) {
//     posts(first: 3) {
//       hasMore
//       items {
//         id
//         title
//       }
//     }
//   }
//
const QUERY_SET_ITEMS_KEY = 'items'
