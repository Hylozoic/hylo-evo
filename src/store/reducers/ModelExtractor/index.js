import { Attribute, ForeignKey, ManyToMany } from 'redux-orm/lib/fields'
import { compact, filter, mapValues } from 'lodash'
import { isUndefined, omitBy } from 'lodash/fp'

export default class ModelExtractor {
  static addAll ({ session, root, modelName, ...opts }) {
    if (!root) return

    const extractor = new ModelExtractor(session, opts)
    extractor.walk(root, modelName)
    extractor.addAll()
  }

  constructor (session, options = {}) {
    this.session = session
    this.options = options
    this.accumulator = []
  }

  addAll () {
    const method = this.options.append ? 'updateAppending' : 'update'
    const { isHoloData } = this.options
    this.mergedNodes().forEach(({ modelName, payload }) => {
      const model = this.session[modelName]
      const sourceTaggedPayload = isHoloData
        ? {...payload, isHoloData}
        : {...payload, isHyloData: true}
      model.hasId(sourceTaggedPayload.id)
        ? model.withId(sourceTaggedPayload.id)[method](sourceTaggedPayload)
        : model.create(sourceTaggedPayload)
    })
  }

  walk (node, modelName, atRoot = true) {
    if (atRoot) {
      if (node.hasOwnProperty(QUERY_SET_ITEMS_KEY) || Array.isArray(node)) {
        return this._walkMany(node, modelName)
      }
    }

    const model = this.session[modelName]
    if (!model) {
      throw new Error(`no model in session named "${modelName}"`)
    }

    const normalized = omitBy(isUndefined, mapValues(node, (value, key) => {
      var type = model.fields[key]

      if (value && value.__typename) {
        const polymorphicChildId = this._walkOne(value, value.__typename)
        return `${value.__typename}-${polymorphicChildId}`
      }

      if (type instanceof Attribute) {
        return value
      }

      if (type instanceof ForeignKey) {
        return this._walkOne(value, type.toModelName)
      }

      if (type instanceof ManyToMany) {
        return this._walkMany(value, type.toModelName)
      }

      if (!type && key in model.prototype) {
        // this is a reverse relation defined by relatedName
        type = model.virtualFields[key]

        if (type instanceof ForeignKey) {
          // each of the related values needs to have a foreign key back to
          // the current value...
          this._walkMany(value, type.toModelName, {[type.relatedName]: node.id})

          // ...and because the related values store the foreign key, the
          // current value does not need to record anything about the relation,
          // so we return nothing
          return
        }

        if (type instanceof ManyToMany) {
          return this._walkMany(value, type.toModelName)
        }
      }

      if (!type) return value
      throw new Error(`don't know how to handle type: ${type}`)
    }))

    if (normalized.id) {
      this.accumulator.push({
        modelName: model.modelName,
        payload: normalized
      })
    }
  }

  mergedNodes () {
    return mergeDuplicates(this.accumulator)
  }

  _walkOne (value, modelName) {
    if (typeof value !== 'object' || !value) return value
    this.walk(value, modelName, false)
    return value.id
  }

  _walkMany (value, modelName, extraProps) {
    const items = Array.isArray(value) ? value : value[QUERY_SET_ITEMS_KEY]
    return items.map(x => {
      const node = extraProps ? Object.assign(x, extraProps) : x
      this.walk(node, modelName, false)
      return x.id
    })
  }
}

// This is more complicated than a simple groupBy(modelName + id).map(merge)
// because we need to preserve the order of nodes.
function mergeDuplicates (nodes) {
  const usedIndexes = {}

  // go through all nodes we accumulated from beginning to end and merge
  // duplicates for each one.
  return compact(nodes.map((node, index) => {
    // skip this node if it's already been merged with an earlier one
    if (usedIndexes[index]) return
    const { modelName, payload: { id } } = node

    // find nodes later in the list that have the same model name and ID,
    // collect them, and mark them to be skipped later
    const duplicates = filter(nodes, (other, index2) => {
      if (index2 > index &&
        other.modelName === modelName && other.payload.id === id) {
        usedIndexes[index2] = true
        return true
      }
    })

    if (duplicates.length === 0) return node

    // merge all the payloads into a new one.
    // we don't use lodash.merge here because it merges arrays by
    // concatenating them, and we don't need its "deep" behavior since these
    // nodes are already normalized.
    const payloads = [{}, node.payload].concat(duplicates.map(d => d.payload))
    return {
      modelName,
      payload: Object.assign.apply(null, payloads)
    }
  }))
}

// this is the key under which our GraphQL API stores the list of items for
// "query-set style" pagination, e.g.:
//
//   person(id: 2) {
//     posts(first: 3) {
//       hasMore
//       items {
//         id
//         title
//       }
//     }
//   }
//
const QUERY_SET_ITEMS_KEY = 'items'
