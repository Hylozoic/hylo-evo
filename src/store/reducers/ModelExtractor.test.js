import testPayloads from './ModelExtractor.test.json'
import ModelExtractor from './ModelExtractor'
import orm from '../models'

const payload = testPayloads['FETCH_POSTS for me']
const payload2 = testPayloads['FETCH_POSTS for community']

it('produces a flat list from a "separate-totals-style" response', () => {
  const extractor = new ModelExtractor(orm.session(orm.getEmptyState()))
  extractor.walk(payload.data.me, 'Me')
  expect(extractor.mergedNodes()).toMatchSnapshot()
})

it('produces a flat list from a "nested-totals-style" response', () => {
  const extractor = new ModelExtractor(orm.session(orm.getEmptyState()))
  extractor.walk(payload2.data.community, 'Community')
  expect(extractor.mergedNodes()).toMatchSnapshot()
})
