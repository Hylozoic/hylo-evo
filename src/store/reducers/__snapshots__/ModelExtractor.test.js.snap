// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`produces a flat list from a "nested-totals-style" response 1`] = `
Array [
  Object {
    "modelName": "Community",
    "payload": Object {
      "avatarUrl": "http://foo.com/avatar.png",
      "bannerUrl": "http://foo.com/banner.png",
      "id": "1",
      "memberCount": 42,
      "name": "foo",
      "posts": Array [
        "2",
        "3",
      ],
      "slug": "foo",
    },
  },
  Object {
    "modelName": "Post",
    "payload": Object {
      "communities": Array [
        "1",
      ],
      "id": "2",
      "title": "Hello!",
    },
  },
  Object {
    "modelName": "Post",
    "payload": Object {
      "communities": Array [
        "1",
      ],
      "id": "3",
      "title": "Hello Again!",
    },
  },
]
`;

exports[`produces a flat list from a "separate-totals-style" response 1`] = `
Array [
  Object {
    "modelName": "Person",
    "payload": Object {
      "avatarUrl": "https://www.gravatar.com/avatar/32f7ff34c3b5a8600b79be9a85a5c92f?d=mm&s=140",
      "bannerUrl": "https://pbs.twimg.com/profile_banners/18696471/1435442798/1500x500",
      "id": "46816",
      "name": "Rich",
    },
  },
  Object {
    "modelName": "Person",
    "payload": Object {
      "avatarUrl": "https://www.gravatar.com/avatar/32f7ff34c3b5a8600b79be9a85a5c92f?d=mm&s=140",
      "id": "99",
      "name": "Flargle Argleson",
    },
  },
  Object {
    "modelName": "Community",
    "payload": Object {
      "id": "1836",
      "members": Array [
        "99",
      ],
      "name": "Wombats",
    },
  },
  Object {
    "modelName": "Comment",
    "payload": Object {
      "createdAt": "Fri Mar 24 2017 10:55:41 GMT+1300 (NZDT)",
      "creator": "46816",
      "id": "1",
      "post": "30002",
      "text": "<p>Here's a comment!</p>",
    },
  },
  Object {
    "modelName": "Post",
    "payload": Object {
      "commentsTotal": "1",
      "communities": Array [
        "1836",
      ],
      "communitiesTotal": "1",
      "createdAt": "Sat Mar 18 2017 10:48:43 GMT+1300 (NZDT)",
      "creator": "46816",
      "details": "<p><a href=\\"https://wombat.life\\">https://wombat.life</a></p>
<p></p>
<p></p>",
      "endsAt": null,
      "followers": Array [
        "46816",
      ],
      "followersTotal": "1",
      "fulfilledAt": null,
      "id": "30002",
      "startsAt": null,
      "title": "Hello",
      "type": null,
    },
  },
  Object {
    "modelName": "Comment",
    "payload": Object {
      "createdAt": "Fri Mar 24 2017 10:50:41 GMT+1300 (NZDT)",
      "creator": "46816",
      "id": "2",
      "post": "30003",
      "text": "<p>Another comment.</p>",
    },
  },
  Object {
    "modelName": "Post",
    "payload": Object {
      "commentsTotal": "1",
      "communities": Array [
        "1836",
      ],
      "communitiesTotal": "1",
      "createdAt": "Sat Mar 18 2017 10:49:43 GMT+1300 (NZDT)",
      "creator": "46816",
      "details": "<p>So true.</p>",
      "endsAt": null,
      "followers": Array [
        "46816",
      ],
      "followersTotal": "1",
      "fulfilledAt": null,
      "id": "30003",
      "startsAt": null,
      "title": "Yup",
      "type": null,
    },
  },
  Object {
    "modelName": "Me",
    "payload": Object {
      "avatarUrl": "http://foo.com/bar.png",
      "id": "1",
      "name": "Foo Bar",
      "posts": Array [
        "30002",
        "30003",
      ],
    },
  },
]
`;
